Tutorial ABCD method with RooParamtricHist

cesarecazzaniga opened the pull request:
Pull request to propose a small tutorial to illustrate the application of RooParametricHist for a per-bin ABCD method in Combine.

Material added in the PR:

    documentation for the tutorial
    codes to generate input datasets, workspaces and cards to run the tutorial

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,468 @@
+To generate your own input data, run: 
+
+```
+python utils/produce_input_histograms_and_analyse.py
````
updated with python3 command in instructions

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,468 @@
+To run the workspace creation script:
+
+```
+python utils/create_workspace.py -m 1500
````
i have harmonized where files are saved and fetched in the code such that this is automatised and users have not to specify the path

cesarecazzaniga commented on data/tutorials/abcd_rooparametrichist_exercise/utils/create_workspace.py:
```diff
@@ -0,0 +1,248 @@
+    print ("Reading histogram: ", hist_nameC)
+    print ("Reading histogram: ", hist_nameD)
+    histA = input_file.Get(hist_nameA)
+    histA.SetDirectory(0)
```
solved issue

kcormi replied:
I'm still seeing the same issue. I don't see a change in either the name of the root histograms that are produced in the previous step or the ones that are checked for here. Did I miss an update, or maybe it didn't get committed?

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,468 @@
+
+
+```
+python utils/create_datacards.py -m 1500
````
done

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,468 @@
+
+```
+
+The datacards can be combined then using the usual command:
````
i have added commands

kcormi replied:
I don't see these added either.  Are they pushed?

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,468 @@
+
+```
+
+Using the output ```higgsCombineTest.FitDiagnostics.mH1500.root```, one can run the script ```$CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/test/mlfitNormsToText.py``` to get the predictions for the normalizations.
````
added in the tutorial text

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,468 @@
+Using the output ```higgsCombineTest.FitDiagnostics.mH1500.root```, one can run the script ```$CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/test/mlfitNormsToText.py``` to get the predictions for the normalizations.
+
+<details>
+<summary> Normalizations predictions </summary>
````
solved issue, added new datacards and new code to produce them

kcormi replied:
Thanks, yes, I now get the same values for pre-fit, and my S+B fit values were slightly different, but perhaps thats just due to the initialized data?

cesarecazzaniga commented:
I went through your comments, and i have modified the tutorial accordingly. First, in the generation of the files i have fixed the seed such that we should all get the same results. Second, i have noticed a small issue in the datacards generation code that was creating the large r fitted value you were noticing, and i fixed it. I have added the results one should get to the tutorial and the new plots. Moreover i have added, as suggested, the commands to really reproduce each step.

kcormi commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,507 @@
+From the combined datacard for all the regions, one can run the usual fit diagnostics as follows:
+
+```
+combine -M FitDiagnostics combined_mPhi_1500_2018.txt -m 1500 --saveShapes --saveWithUncertainties --saveNormalizations
````
Here, I am now getting an issue that the datacard seems to be looking for `mPhi_{mass}_{region}` but the relevant histograms in my workspace appear to be saved as `mPhi{mass}_{region}`.

cesarecazzaniga replied:
this should be solved now since i forgot to push the new version of the script creating the workspace

kcormi commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,507 @@
+</details>
+
+
+Using the output ```fitDiagnosticsTest.mH1500.root```, one can run the script ```$CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/test/mlfitNormsToText.py``` to get the predictions for the normalizations:
````
The output file in this case does not include the mass parameter, its just `fitDiagnosticsTest.root`

cesarecazzaniga replied:
this is strange, since i get exactly that outfile name running this command: ```combine -M FitDiagnostics combined_mPhi_1500_2018.txt -m 1500 --saveShapes --saveWithUncertainties --saveNormalizations```. I would expect the ```.mH1500.``` part since the mass parameter is specified

cesarecazzaniga replied:
i have updated, now is consistent with what you also get in terms of filename (was running with an older version)

kcormi replied:
Are you sure? The filenames produced for the files which contain the limit tree (which do always contain the mass parameter, but is not relevant here) do not follow the same output formatting as the fitDiagnostic file which saves the shapes and RooFitResults. The `higgsCombineTest.FitDiagnostics.mH1500.root` file should also be produced but that just contains the limit tree, the `fitDiagnosticsTest.root` file doesn't rely on the mass value and only uses the `-n` argument.

kcormi commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,507 @@
+Moreover, you can run the script in ```$CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/data/tutorials/longexercise/postFitPlot.py``` to get pre-fit and post-fit plots in the signal region (in the combined datacard ```ch4```):
+
+```
+python3 $CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/data/tutorials/longexercise/postFitPlot.py --input_file fitDiagnosticsTest.mH1500.root --shape_type <shapes_type> --region <region>
````
Same filename issue as above.

cesarecazzaniga replied:
now changed, i was running with a slightly older version of combine where the mass was added in the name of the file

kcormi reviewed (commented):
Thanks Cesare! Some of the issues are fixed, but I still had a few issues when running through the tutorial to be fixed.

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,507 @@
+## Generate input data
+<a id="inputs"></a>
+
+The histograms for the $z$ observable in the different regions A,B,C,D can be produced using the ```produce_input_histograms_and_analyse.py``` script in ```utils/produce_input_histograms_and_analyse.py```. In the script the expected rates for different signal hypotheses (as a function of $\Phi$ mass $m_{\Phi} \in \{1500, 2000, 3000, 4000, 5000 \}$ GeV) and the background yields are specified, as  ...
````
done

cesarecazzaniga commented:
added uncommitted changes for create-workspace.py, updated documentation to match naming of files from most recent version of combine, updating results to match the ones from the latest version

kcormi commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,519 @@
+
+```
+
+The datacards will be created in the directory ```example_analysis/datacards/mPhi1500/``` inside the tutorial directory. The datacards can be combined then using the usual command:
````
Can we just make sure to add a statement to go change into this directory (or add the cd command below, or both).

cesarecazzaniga replied:
i have updated the text as you suggested by adding explicit instruction to enter the directory

kcormi reviewed (commented):
Thanks Cesare! This is working for me now. I have one remaining minor comment. After that, I think this can be merged unless any one else has some comments.

It might be nice at some point to expand this to e.g. show how to setup different mass points with a single datacard using keywords. But I don't think that those kinds of future developments should stand in the way of including this as is for now.
